src/github.com/matrix-org/dendrite/syncapi/storage/syncserver.go

```diff
@@ -594,18 +597,29 @@ func (d *SyncServerDatabase) getStateDeltas(
 	for _, joinedRoomID := range joinedRoomIDs {
 		deltas = append(deltas, stateDelta{
 			membership:  "join",
-			stateEvents: streamEventsToEvents(state[joinedRoomID]),
+			stateEvents: streamEventsToEvents(device, state[joinedRoomID]),
 			roomID:      joinedRoomID,
 		})
 	}
 
 	return deltas, nil
 }
 
-func streamEventsToEvents(in []streamEvent) []gomatrixserverlib.Event {
+func streamEventsToEvents(device *authtypes.Device, in []streamEvent) []gomatrixserverlib.Event {
 	out := make([]gomatrixserverlib.Event, len(in))
 	for i := 0; i < len(in); i++ {
 		out[i] = in[i].Event
+		if device != nil && in[i].transactionID != nil {
+			if device.ID == in[i].transactionID.DeviceID {
```

Comment:
do we not need to check the userid, given you've told me that DeviceIDs are not unique?